pkg/plugins/backendplugin/instrumentation/instrumentation.go

```diff
@@ -58,29 +58,105 @@ const (
 	endpointQueryData      = "queryData"
 )
 
+type errorSource string
+
+const (
+	grafanaSource          errorSource = "grafana"
+	userSource             errorSource = "user"
+	downstreamClientSource errorSource = "downstreamClient"
+	downstreamServerSource errorSource = "downstreamServer"
+	noneSource             errorSource = "none"
+)
+
 var logger = plog.New("plugin.instrumentation")
 
 // instrumentPluginRequest instruments success rate and latency of `fn`
 func instrumentPluginRequest(ctx context.Context, cfg Cfg, pluginCtx *backend.PluginContext, endpoint string, fn func() error) error {
-	status := statusOK
-
 	start := time.Now()
 	timeBeforePluginRequest := log.TimeSinceStart(ctx, start)
 
 	err := fn()
-	if err != nil {
-		status = statusError
-		if errors.Is(err, context.Canceled) {
-			status = statusCancelled
+
+	elapsed := time.Since(start)
+	status := getStatus(err)
+
+	logger.Info("Plugin Request Completed")
+
+	updateMetrics(pluginCtx.PluginID, endpoint, string(cfg.Target), elapsed, status, "none")
+	logDatasourceRequests(ctx, cfg, pluginCtx, endpoint, status, elapsed, timeBeforePluginRequest, err)
+
+	return err
+}
+
+func InstrumentQueryDataRequest(ctx context.Context, cfg Cfg, pluginCtx *backend.PluginContext, requestSize float64, fn func() (*backend.QueryDataResponse, error)) (*backend.QueryDataResponse, error) {
+	start := time.Now()
+	timeBeforePluginRequest := log.TimeSinceStart(ctx, start)
+
+	resp, err := fn()
+
+	elapsed := time.Since(start)
+	logger.Info("Plugin Request Completed")
+
+	status := getStatus(err)
+	errorSource := getErrorSource(status, resp)
+
+	updateMetrics(pluginCtx.PluginID, endpointQueryData, string(cfg.Target), elapsed, status, string(errorSource))
+	pluginRequestSizeHistogram.WithLabelValues("grafana-backend", pluginCtx.PluginID, endpointQueryData, string(cfg.Target)).Observe(requestSize)
+	logDatasourceRequests(ctx, cfg, pluginCtx, endpointQueryData, status, elapsed, timeBeforePluginRequest, err)
+
+	return resp, err
+}
+
+func getErrorSource(status string, resp *backend.QueryDataResponse) errorSource {
+	if status == statusError {
+		return grafanaSource
```

Comment:
Couldn't this be either Grafana or downstream?

Reply:
it all depends of the plugins implementation I guess. For Prometheus, if this return an error, then it is something that happened before the request to the downstream. Then whatever the downstream response status code, no error are returned.
In that case it seems like if there is an error, then it is before downstream is involved.

Different plugins can have different implementation and behavior regarding when they return error or not.

The plugin-sdk interfaces seems to be limited here: error/ not error. 
Ideally we could pass some free form meta data too, so we can keep the current behavior for error, but let plugins returns meta data about where the error code from for example

Reply:
I think it would be reasonable to add something to the plugin-sdk about where the error came from. It's fine if it does not work for all current plugins. The goal here is to improve the monitoring for the datasources we care the most about and those would be updated quickly. 

@marefr WDYT?

Reply:
@bergquist I'm not sure that make sense. We already allow plugins to define their own metrics, wonder if that being the way forward given different plugins might have different needs. Do you have any suggestion of what that might look like?

Reply:
Think my main concern with all this is that we would have more errorSources than just Grafana and downstream/plugin. From the perspective of defining SLO's for a team owning a plugin my thinking is that Grafana and downstream/plugin should be enough to understand if something is wrong with a certain plugin. But I guess the problem is that SLO's for those owning the plugin need to exclude errors happening downstream - so basically the same problem Grafana Cloud operators have today :exploding_head: 

Based on this I've come to the conclusion that each plugin should define their own custom metrics to be used for SLO purposes. Eager to hear your thoughts regarding this